Fix mysql article id error

RajPrakash681 opened the pull request:
## Problem Description
Fixes #6470 - CourseDataUpdateWorker was failing with `Mysql2::Error: Field 'article_id' doesn't have a default value` in MySQL strict mode when processing timeslices.

### Root Cause
The error occurs when `RevisionOnMemory` objects are created with `nil` article_id because the `mw_page_id` cannot be found in the articles dictionary during revision processing. This specifically affected the Florida_State_University/Afro-diasporic_Legacies_(Fall) course timeslice from 2025-10-03 to 2025-10-04.


## Files Modified
- `app/models/article_course_timeslice.rb`: Added validation and error handling
- `app/services/update_course_wiki_timeslices.rb`: Added filtering and exception handling  
- `lib/revision_data_manager.rb`: Added debug logging
- `spec/models/article_course_timeslice_spec.rb`: Added comprehensive tests
- `spec/services/update_course_wiki_timeslices_spec.rb`: Added edge case tests
- `spec/factories/article_course_timeslices.rb`: Fixed incomplete factory
 




The multi-layer approach ensures the system gracefully handles cases where article_id is nil while maintaining data integrity and providing comprehensive logging for debugging future issues.

Abishekcs commented on app/models/article_course_timeslice.rb:
```diff
@@ -62,12 +74,19 @@ def self.update_article_course_timeslices(course, article_id, revisions)
+                                                                start: timeslice.start,
+                                                                end: timeslice.end)
+        # Update cache for ArticleCourseTimeslice
+        ac_timeslice.update_cache_from_revisions revisions_in_timeslice
```
Better use brackets

Abishekcs commented on app/services/update_course_wiki_timeslices.rb:
```diff
@@ -181,13 +181,34 @@ def update_article_course_timeslices_for_wiki(revisions)
+      begin
+        ArticleCourseTimeslice.update_article_course_timeslices(@course, article_id,
+                                                                article_revisions_data)
+      rescue StandardError => e
```
Same for this since you are already filtering out `rev` with `nil article id`

Abishekcs commented on lib/revision_data_manager.rb:
```diff
@@ -164,11 +164,22 @@ def mark_scoped_articles(wiki, revisions)
+    
+    # Log a warning if we can't find the article_id for a revision
+    # This helps with debugging the original issue #6470
+    if !articles.nil? && article_id.nil?
```
I think only `article_id.nil?` should be enough.Am I missing something?

RajPrakash681 commented:
> > > @RajPrakash681 Where you able to reproduce this bug locally?
> > 
> > 
> > @Abishekcs I reproduced this locally on my development machine running MySQL 8.0 with strict mode enabled. The error occurred when running course updates for courses that had revisions referencing mw_page_id values missing from our local articles table.
> > I triggered it by running UpdateCourseWikiTimeslices on a course with stale article data, which caused find_or_create_by to attempt creating records with article_id: nil, resulting in MySQL's "Field 'article_id' doesn't have a default value" error.
> 
> Interesting! I would like to understand what "a course with stale article data" exactly means. Do you mind sharing the steps you followed to reproduce the error? Importing article records is a side effect of fetching revisions (see the `fetch_revision_data_for_course` definition), so it's hard for me to see how those articles are imported but then, when the revisions in memory are created, article_id is nil. Maybe it's related to resolving duplicates (see the `resolve_duplicate_articles` method)?
> 
> Anyway, my intuition says that the approach to fix this problem should be more about avoiding the creation of `RevisionOnMemory` objects with a nil `article_id` than about handling those nil IDs when processing timeslices.

Thanks for the great feedback! You’re right , the fix should focus on preventing nil article_id values, not handling them later. “Stale article data” refers to revisions pointing to articles marked as deleted: true during resolve_duplicate_articles.

Flow:

1.fetch_revision_data_for_course imports revision data.
2.Queries for deleted: false articles.
3.resolve_duplicate_articles may mark some as deleted.
4.The article dictionary is built from stale data.
5.Deleted article lookups return nil.
and now i have updated the fix as you said !

RajPrakash681 commented:
@gabina please have a look now!
